### ISSUE TITLE
False positive on "consider deriving a `Default` implementation"

### ISSUE THREAD

**Issue by tarcieri** [Manishearth/rust-clippy#1840]
With clippy 0.0.140. Example code:

```rust
pub trait Herp {}

#[derive(Default)]
pub struct Derp<K: Herp> {
    table: Vec<K>
}

impl<K> Derp<K> where K: Herp {
    pub fn new() -> Derp<K> {
        Derp { table: Vec::new() }
    }
}
```

With this, I get the following clippy warning/error:

```
error: you should consider deriving a `Default` implementation for `derp::Derp<K>`
  --> src/derp.rs:9:5
   |
9  | /     pub fn new() -> Derp<K> {
10 | |     Derp { table: Vec::new() }
11 | |   }
   | |___^
   |
   = note: #[deny(new_without_default_derive)] implied by #[deny(warnings)]
note: lint level defined here
  --> src/lib.rs:9:61
   |
9  | #![deny(missing_docs, trivial_casts, trivial_numeric_casts, warnings)]
   |                                                             ^^^^^^^^
   = help: for further information visit https://github.com/Manishearth/rust-clippy/wiki#new_without_default_derive
help: try this
   | #[derive(Default)]
```

**Comment by Manishearth** [comment]
Ah, this is because we use `implements_trait`, but `Derp<K>` (i.e. `Self`) doesn't implement default, `Derp<K: Default>` does.

Perhaps we should use the defid to search for all trait impls and filter them out? This will be slow so we should only do this after checking `implements_trait`

**Comment by Manishearth** [comment]
```
﻿11:59 <&Manishearth> eddyb: is there an easy way to check if Foo<T> implements a trait for *any* T?
12:00 < eddyb> Manishearth: trivial
12:00 < eddyb> Manishearth: just leave it a TyParam and don't have bounds involving it in ParamEnv
12:00 < eddyb> it's easier than checking in a given context :P
12:00 < eddyb> usually
12:01 < eddyb> Manishearth: have to use a fulfillment context but you always need that I think
```
